src/main/java/com/example/gymApp/repository/TrainingRepository.java

```diff
@@ -0,0 +1,50 @@
+package com.example.gymApp.repository;
+
+import com.example.gymApp.model.Trainee;
+import com.example.gymApp.model.Trainer;
+import com.example.gymApp.model.Training;
+import java.time.LocalDate;
+import java.util.List;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface TrainingRepository extends JpaRepository<Training, Long> {
+
+
+
+  @Query("SELECT t FROM Training t WHERE t.trainee.user.username = :traineeUsername "
+      + "AND t.trainer.user.firstName = :trainerName "
+      + "AND t.trainer.specialization.name = :specialization "
+      + "AND t.trainingDate BETWEEN :startDate AND :finishDate")
+  List<Training> getAllTrainingsByTraineeAndCriteria(
```

Comment:
Criteria are usually used as a filter and refine the main search. This means they are not required. 
Have you implemented partial search by username?